feat: add Streamlit UI with full configuration

chrisgscott opened the pull request:
## Streamlit UI for Piragi

Interactive chat interface that demonstrates all of Piragi's features.

### Features

**Document Management**
- 📁 Upload files (txt, md, pdf, html, docx) or add URLs
- 💾 Persistent storage in `~/.piragi/uploads/`
- 📂 View saved documents
- �� Re-index with new settings
- 🗑️ Delete everything option

**Chat Interface**
- 💬 Grounded Q&A with citations
- 📎 Expandable sources with relevance scores
- 🔍 Debug mode showing timing and config

**Full Configuration**
- **LLM**: Model selection (Ollama)
- **Chunking**: Strategy (fixed/semantic/hierarchical/contextual) with strategy-specific params
- **Retrieval**: HyDE, Hybrid Search, Reranker toggles
- **Top K**: Control result count

### Quick Start

```bash
./ui/run.sh
```

The script checks for Ollama, installs dependencies, and launches the UI.

### Screenshots

The UI provides:
1. Settings panel with all chunking/retrieval options
2. Document upload and management
3. Chat with grounded answers and citations

### 100% Local

Runs entirely on your machine with Ollama. No external API calls for embeddings or LLM.

### Files

- `ui/app.py` - Main Streamlit application (~380 lines)
- `ui/run.sh` - Quick start script
- `ui/README.md` - Documentation

hemanth commented:
Should we convert it into a webapp? 🤔

hemanth commented:
Maybe have both? The advantage of a hosted app is users can get a preview of how piragi behaves with a single click.

chrisgscott commented:
**Updated with HF Spaces demo app**

Added [demo_app.py](cci:7://file:///Users/chrisgscott/projects/piragi/ui/demo_app.py:0:0-0:0) - a hosted-friendly variant for Hugging Face Spaces:

**Demo Features:**
- **Session isolation** - Each user gets their own ephemeral storage (UUID-based temp dirs)
- **API flexibility** - Works with HF Inference API, OpenAI, or falls back to local Ollama
- **Sample docs** - Pre-loads piragi README & API.md so users can try immediately
- **Example questions** - Clickable prompts to help new users get started
- **Learn More** - Sidebar expander explaining chunking strategies and retrieval options

**Bug Fixes (both apps):**
- Citations now sorted by relevance (highest first)
- Fixed negative score display from distance-based retrieval
- KB auto-recreates when chunking config changes (was using stale config)

**Files:**
- [ui/demo_app.py](cci:7://file:///Users/chrisgscott/projects/piragi/ui/demo_app.py:0:0-0:0) - New demo app
- `ui/requirements-demo.txt` - HF Spaces dependencies
- [ui/app.py](cci:7://file:///Users/chrisgscott/projects/piragi/ui/app.py:0:0-0:0) - Citation sorting fix
- `ui/README.md` - Demo instructions

**Deployment Note:**
For HF Spaces, set `HF_TOKEN` as a Space secret (or `OPENAI_API_KEY` if using OpenAI). The app auto-detects which API to use based on available env vars. Without either, it falls back to local Ollama.

---

Also opened a separate PR (#6) for a LanceDB score normalization bug we found during testing - scores can go negative when L2 distance > 1.0.
